[WIP] Porting Zephyr to Unisoc uwp5661 chip.

easonxiang opened the pull request:
Unisoc UWP5661 is a highly integrated connectivity single chip which offers the lowest BOM in the industry for smart phone, PC, STB, OTT, IoT and automotive applications.
This chip includes 2.4GHz and 5GHz WLAN IEEE 802.11 a/b/g/n/ac, Bluetooth 5.0 with supporting high power mode, Direction Finding and Long Range.

This patch is initial commit, which include uwp5661 soc support, hal library, watchdog and uart driver.

galak commented:
Thanks for the submission.  Please review http://docs.zephyrproject.org/contribute/contribute_guidelines.html  and http://docs.zephyrproject.org/contribute/contribute_non-apache.html

1. Commit messages are required in each commit
2. Imports to EXT need a clear license statement and README.

Thanks

galak commented:
> @galak Thanks for the review, patches have been updated as requirement. EXT files are under apache2.0 license. Please comment if hvae any problem.
> 
> Thanks.

Can you update all src files to have a SPDX tag in them to be clear on the license
```
/*
 * Copyright (c) 2018 Linaro Limited.
 *
 * SPDX-License-Identifier: Apache-2.0
 */
```

galak commented:
> @galak Thanks for the review, patches have been updated as requirement. EXT files are under apache2.0 license. Please comment if hvae any problem.

How does one get the HAL/SDK for the Unisoc UWP360?

easonxiang commented:
Unisoc UWP360 HAL/SDK is under development, more module interfaces/drivers will be added after various tests. The full SDK will be released once it was ready.

ioannisg commented on arch/arm/soc/unisoc_uwp/uwp3/Kconfig.series:
```diff
@@ -0,0 +1,11 @@
+
+config  SOC_SERIES_UWP3
+	bool "Unisoc UWP3"
+	select CPU_CORTEX_M
```
Not needed to specify CPU_CORTEX_M, if you have already selected CPU_CORTEX_M4

easonxiang replied:
Thanks for the review, patch has been updated.

SebastianBoe reviewed (changes requested):
A board .rst file that includes information like where to get such a board must be included.

galak commented:
@easonxiang any updates on the availability of the HAL/SDK?  and on updating this PR.

easonxiang commented:
@SebastianBoe The Soc Chip and Board are not announced yet, the document for zephyr is no available currently, I will update it when it was ready.

galak commented:
Any updates on the availability of the SDK/HAL from a public website?

easonxiang commented:
There is not. Actually, Github is our first commit to the public website.

Currently, we are debugging Wi-Fi functionality, scan and connect are works on Zephyr system.

pfalcon commented:
@ioannisg : As your comment seems to have been addressed, perhaps you could dismiss your -1.

pfalcon commented:
@easonxiang : This now has conflicts, can you please rebase on master?

easonxiang commented:
@pfalcon Thanks for your comment, I wanna to know if it is better to rebase on 1.13 release as  there are so many changes on master and our patches?

easonxiang commented:
In addition, we are preparing a demo for the coming Linaro Connect, so we may rebase on the latest code after the Connect.

pfalcon commented:
@easonxiang:

> I wanna to know if it is better to rebase on 1.13 release as there are so many changes on master and our patches?

I would suggest to rebase regularly, e.g. at least once a week. This way you won't accumulate conflicts and discrepancies with mainline, but resolve them quickly one by one as they appear. Of course, that might not work for every workflow, so up to you.

> we are preparing a demo for the coming Linaro Connect

Cool, looking forward to it.

Commit:
soc: arm: Add new SOC Unisoc uwp5661 support.

Unisoc UWP5661 is a highly integrated connectivity single chip.

This chip includes 2.4GHz and 5GHz WLAN IEEE 802.11 a/b/g/n/ac,
Bluetooth 5.0 with supporting high power mode, Direction Finding
and Long Range.

Signed-off-by: Dong Xiang <dong.xiang@unisoc.com>

Commit:
dts: Add dts files for Unisoc uwp5661 chip.

dts files for Unisoc uwp5661 chip.

Signed-off-by: Dong Xiang <dong.xiang@unisoc.com>

Commit:
ext: hal: Add HAL library for Unisoc uwp5661 chip.

uwp5661 HAL library, provide peripheral device interfaces to soc
and drivers.

Signed-off-by: Dong Xiang <dong.xiang@unisoc.com>

Commit:
drivers: watchdog: Add watchdog support for Unisoc uwp5661 chip.

uwp5661 onchip watchdog, support RESET, INTERRUPT and MIX mode.

Signed-off-by: Dong Xiang <dong.xiang@unisoc.com>

Commit:
drivers: serial: Add serial support for Unisoc uwp5661 chip.

Enable uwp5661 onchip uart.

Signed-off-by: Dong Xiang <dong.xiang@unisoc.com>

Commit:
board: arm: Add IVY5661 96Boards support.

IVY5661 96Boards based on Unisoc uwp5661 chip.

Signed-off-by: Dong Xiang <dong.xiang@unisoc.com>

galak commented:
Can we rebase this PR.  Also I understand that the SoC/chip got renamed, so we should probably make that change as well.

Commit:
drivers: interrupt_controller: Add INTC driver for uwp5661.

add external interrupt controller support.

Signed-off-by: Dong Xiang <dong.xiang@unisoc.com>

easonxiang commented:
@galak Thanks for your comment.
This PR has been rebased on the latest code, and chip name was renamed to uwp5661.

dbkinder commented on boards/arm/96b_ivy5661/doc/96b_ivy5661.rst:
```diff
@@ -0,0 +1,112 @@
+
+This board support Wi-Fi 802.11ac 2x2 dual band and Bluetooth 5 with
+high power mode. It can be used by intelligent household, IP camera,
+Wi-Fi repeater and other networking applications.
```
add a comma:  ``repeater, and``

easonxiang replied:
Done.

dbkinder reviewed (changes requested):
doc look fine, with a couple of little tweaks please.

easonxiang commented on boards/arm/96b_ivy5661/doc/96b_ivy5661.rst:
```diff
@@ -0,0 +1,112 @@
++-----------+------------+-------------------------------------+
+
+
+The default board configuration can be found in the defconfig file:
```
>This renders on the website better as:
>
>The default board configuration can be found in the defconfig file:
>``boards/arm/96b_ivy5661/96b_ivy5661_defconfig``

Thanks for the review.
Do you mean remove the blank line between these two lines?

easonxiang replied:
done.

galak commented on boards/arm/96b_ivy5661/board.h:
```diff
@@ -0,0 +1,26 @@
+#define __INC_BOARD_H
+
+/* Push button switch 2 */
+#define SW2_GPIO_PIN	6    /* GPIO22/Pin15 */
```
We should move this into the device tree, take a look at this commit to see an example:

commit 619049a74c8b7179ca4925f243597359e4930640
Author: Kumar Gala <kumar.gala@linaro.org>
Date:   Fri Sep 28 16:35:16 2018 -0500

    boards: nrf51_ble400: Convert to using dts for led/button

easonxiang replied:
This modification will take some times.

galak commented on boards/arm/96b_ivy5661/dbghdr.c:
```diff
@@ -0,0 +1,14 @@
+
+#ifdef BOARD_96B_IVY5661
+__attribute__ ((section(".dbghdr")))
+const unsigned long ulDebugHeader[] = {
```
Is this a board specific header, or something we'd need for this SoC always?

easonxiang replied:
This is not used by the SoC, remove this file.

galak commented on boards/arm/96b_ivy5661/support/openocd.cfg:
```diff
@@ -0,0 +1,12 @@
+source [find board/stm32f4discovery.cfg]
```
Is this correct, saying the board cfg is stm32f4discovery.cfg seems wrong

easonxiang replied:
Yes, this it not correct. 
UWP5661 does just support J-Link for debugging. We are still looking for the root cause that how to support OpenOCD as well.
May I remove this config for this commit, and support OpenOCD in the future.

galak commented on drivers/serial/uart_uwp.c:
```diff
@@ -0,0 +1,328 @@
+
+#include "uwp_hal.h"
+#include "uart_uwp.h"
+#include "../interrupt_controller/intc_uwp.h"
```
Is this header needed?

easonxiang replied:
This header file is needed by uart driver.
There is a 2-Level interrupt mechanism in this chip. UART2 interrupt line connect to interrupt controller, and interrupt controller connect to NVIC. Therefor, interrupt controller need to provide interfaces to register uart isr callback, such as uwp_aon_intc_set_irq_callback.

galak commented on drivers/serial/uart_uwp.c:
```diff
@@ -0,0 +1,328 @@
+}
+
+
+	void (*irq_callback_set)(struct device *dev,
```
Why do we need this?

easonxiang replied:
This is a mistake, remove it.

galak commented on dts/arm/unisoc/mem.h:
```diff
@@ -0,0 +1,26 @@
+#ifndef __DT_UNISOC_MEM_H
+#define __DT_UNISOC_MEM_H
+
+#define __SIZE_K(x) (x * 1024)
```
We've moved away from this style of mem.h header.  We introduced DT_SIZE_K/DT_SIZE_M.

easonxiang replied:
done.

galak reviewed (changes requested):
Various cleanups/comments.

easonxiang commented:
Split this patch into 2 parts:

Unisoc HAL, imported from https://github.com/unisoc/hal_uwp5
https://github.com/zephyrproject-rtos/zephyr/pull/10622 

and ivy5661 board support:
https://github.com/zephyrproject-rtos/zephyr/pull/10626
